test/integration/cdi/example-driver/app/kubeletplugin.go

```diff
@@ -0,0 +1,161 @@
+	logger klog.Logger
+	d      plugin.DRAPlugin
+
+	cdiDir     string
+	driverName string
+}
+
+var _ drapbv1.NodeServer = &examplePlugin{}
+
+// runPlugin sets up the plugin and waits for connections from kubelet.
+func runPlugin(logger klog.Logger, cdiDir, driverName, endpoint, draAddress, pluginRegistrationPath string) error {
+	plugin := examplePlugin{
+		logger:     logger,
+		cdiDir:     cdiDir,
+		driverName: driverName,
+	}
+	if err := plugin.start(driverName, endpoint, draAddress, pluginRegistrationPath); err != nil {
+		return fmt.Errorf("start example plugin: %v", err)
+	}
+	// TODO: graceful termination via signal. Needs to be implemented in the caller.
+	select {}
+	plugin.stop()
+	return nil
+}
+
+// getJSONFilePath returns the absolute path where CDI file is/should be.
+func (ex *examplePlugin) getJSONFilePath(claimUID string) string {
+	return filepath.Join(ex.cdiDir, fmt.Sprintf("%s-%s.json", ex.driverName, claimUID))
+}
+
+// start sets up the servers that are necessary for a DRA kubelet plugin.
+func (ex *examplePlugin) start(driverName, endpoint, draAddress, pluginRegistrationPath string) error {
+	// Ensure that directories exist, creating them if necessary. We want
+	// to know early if there is a setup problem that would prevent
+	// creating those directories.
+	if err := os.MkdirAll(ex.cdiDir, os.FileMode(0750)); err != nil {
+		return fmt.Errorf("create CDI directory: %v", err)
+	}
+	if err := os.MkdirAll(filepath.Dir(endpoint), 0750); err != nil {
+		return fmt.Errorf("create socket directory: %v", err)
+	}
+
+	d, err := kubeletplugin.Start(ex.logger, driverName, endpoint, draAddress, pluginRegistrationPath, ex)
+	if err != nil {
+		return fmt.Errorf("start kubelet plugin: %v", err)
+	}
+	ex.d = d
+
+	return nil
+}
+
+// stop ensures that all servers are stopped and resources freed.
+func (ex *examplePlugin) stop() {
+	ex.d.Stop()
+}
+
+// NodePrepareResource ensures that the CDI file for the claim exists. It uses
+// a deterministic name to simplify NodeUnprepareResource (no need to remember
+// or discover the name) and idempotency (when called again, the file simply
+// gets written again).
+func (ex *examplePlugin) NodePrepareResource(ctx context.Context, req *drapbv1.NodePrepareResourceRequest) (*drapbv1.NodePrepareResourceResponse, error) {
+	logger := klog.FromContext(ctx)
+	kind := ex.driverName + "/" + deviceName
+	filePath := ex.getJSONFilePath(req.ClaimUid)
+
+	// Determine environment variables.
+	var env map[string]string
+	if err := json.Unmarshal([]byte(req.ResourceHandle), &env); err != nil {
+		return nil, fmt.Errorf("unmarshal resource handle: %v", err)
+	}
+
+	// CDI wants env variables as set of strings.
+	envs := []string{}
+	for key, val := range env {
+		envs = append(envs, key+"="+val)
+	}
+
+	spec := specs.Spec{
+		Version: cdiVersion,
+		Kind:    kind,
+		Devices: []specs.Device{
+			{
+				Name: deviceName,
+				ContainerEdits: specs.ContainerEdits{
+					Env: envs,
+				},
+			},
+		},
+	}
```

Comment:
You can use top level containerEdits to set Env. The spec will be less nested.

Reply:
But I will still need at least one device, right?

This leads me to my next question: doesn't the device name have to be different for different claims?

Reply:
yes, the device name should be unique for the same kind.